odl/discr/lp_discr.py

```diff
@@ -204,6 +207,11 @@ def element(self, inp=None):
             return self.element_type(self, self.dspace.element())
         elif inp in self.dspace:
             return self.element_type(self, inp)
+        elif isinstance(inp, self.element_type):
+            raise TypeError('input {!r} not an element of {}.'
```

Comment:
The logic is "if in space, ok; otherwise, if `DiscreteLpVector` from foreign space, bad". This keeps vectors from other spaces reaching the sequence-type input part. Maybe one could clarify by writing
```python
elif isinstance(inp, self.element_type) and inp not in space:
    # Same kind, but different space -> error
    raise ...
elif inp in space:
    return ...
```